apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/SessionPool.TargetedSessionPool.cs

```diff
@@ -0,0 +1,602 @@
+                // in the options.
+
+                // In reality, the current pool size should be 0 if pendingAcquisitionCount is non-zero, and
+                // vice versa. However, we don't strictly enforce this, and the maths works out more simply if
+                // we don't worry about it.
+
+                // How many more requests do we need to make in order to get the pool to the minimum size, after satisfying
+                // pending callers, assuming that all the in-flight requests succeed, and there are no more requests?
+                int newRequestsToSatisfyPool = (pendingAcquisitionCount + minPoolSize) - (poolSize + inFlightRequests);
+                // How many more requests *can* we make without going over the maximum number of active sessions?
+                int maxAvailableRequests = maxActiveSessions - (activeSessions + poolSize + inFlightRequests);
+
+                int actualNewRequests = Math.Min(newRequestsToSatisfyPool, maxAvailableRequests);
+                for (int i = 0; i < actualNewRequests; i++)
+                {
+                    Parent.ConsumeBackgroundTask(PrepareNewSessionAsync(), "session creation");
+                }
+
+                async Task PrepareNewSessionAsync()
+                {
+                    try
+                    {
+                        var session = await CreatePooledSessionAsync(CancellationToken.None).ConfigureAwait(false);
+                        ReleaseInactiveSession(session, maybeCreateReadWriteTransaction: true);
+                    }
+                    // Note: we expect this to always actually be an RpcException, but we don't want to end up unhealthy
+                    // with a lot of 
+                    catch (Exception e)
+                    {
+                        Parent._logger.Warn(() => $"Failed to create session for {_databaseName}", e);
+
+                        // CreatePooledSessionAsync will have caused the session pool to become unhealthy already, so we shouldn't get
+                        // any new waiting callers at this point. 
+                        // Propagate this exception to all waiting callers.
+                        List<TaskCompletionSource<PooledSession>> pendingAcquisitionsList = new List<TaskCompletionSource<PooledSession>>();
+                        List<TaskCompletionSource<int>> minimumSizeWaiters;
+                        lock (_lock)
+                        {
+                            // Don't set the exception in the lock.
+                            while (_pendingAcquisitions.TryDequeue(out var acquisition))
+                            {
+                                pendingAcquisitionsList.Add(acquisition);
+                            }
+                            minimumSizeWaiters = _minimumSizeWaiters.ToList();
+                        }
+                        pendingAcquisitionsList.ForEach(tcs => tcs.TrySetException(e));
```

Comment:
I've added a TODO for this, but I think it probably *is* the right approach. It's unlikely that one call would fail but another succeed at roughly the same time, so it's best to fail everything quickly.

If another in-flight call succeeds, that would make the pool healthy again and *future* calls will succeed, but as we don't schedule background tasks for an unhealthy pool, we risk leaving a lot of tasks in limbo if we're not careful.

I think we need to chat about this more :)

Reply:
Maybe having a counter for when the pool has been found unhealthy more than n times in a row or something similar, than could be set/checked in CreatePooledSessionAsync, and when that counter is reached then fault everyone waiting? But agree this might be done later, the current approach is safe so no problem with leaving as is.